### ISSUE TITLE
Des ecriture qui disparaisse tout seul

### ISSUE THREAD

**Issue by pwepwe973** [picsou83/compta.libremen.com#11]
bonjour

J'ai remarqué un bug qui était survenu dans une autre session docker où j'ai dû réinstaller le container complètement. Lorsque j'ouvre une écriture qui existe et que je fais une petite modification, puis que je clique sur "valider", une page "oops écriture inexistante" s'affiche alors que l'écriture existait bien avant de procéder à la modification.

J'ai réinstallé, et là, après avoir passé plusieurs semaines à copier les écritures de mon ancien logiciel vers celui-ci car je n'aimais pas le résultat de l'importation du fichier CSV, le bug resurgit.

Et j'ai peur pour mes écritures car j'ai mis beaucoup de temps à copier tout ça.

Avez-vous déjà remarqué ce bug chez vous ?
![Opera Instantané_2024-04-17_201910_192 168 2 12](https://github.com/picsou83/compta.libremen.com/assets/37281042/5647b705-a763-4024-82ed-e448ef59f9bf)

**Comment by picsou83** [comment]
Chaque écriture possède un numéro unique dans la base de données (id_entry), qui est également associé à un exercice. 

Le logiciel comporte plusieurs sécurités pour empêcher la modification d'une écriture  :

- Lorsque l'exercice est clos.
- Lorsqu'on essaie d'ouvrir directement une écriture avec l'id_entry depuis le mauvais exercice.

Il semble que vous vous trouviez dans ce dernier cas. Vous avez ouvert un onglet avec une écriture (par exemple, depuis l'exercice 2022), puis avez changé d'exercice (par exemple, 2023). Dans ce cas, toute tentative de modification ou visualisation de l'écriture échouera, et ce message d'erreur s'affichera.

Pour travailler sur plusieurs exercices simultanément, il faut ouvrir plusieurs navigateurs et non des onglets.

Si le problème persiste, veuillez le reproduire pour que je puisse l'identifier de mon côté. Cependant, le message "La référence recherchée n'existe pas" ne doit s'afficher que si l'on se trouve dans le mauvais exercice.

**Comment by pwepwe973** [comment]
Je ne saurais pas dire, mais par exemple, si je constate une erreur dans le numéro de pièce et que j'ouvre pour modifier la pièce, eh bien quand je valide, cela me met cette page !

Mais j'ai remarqué que dès que cela m'a fait cela pour une écriture, si j'ouvre les autres pour les modifier, cela me le fera avec tous les autres. Pour remédier à cela, je dois faire une sauvegarde et restaurer les écritures, et là je peux de nouveau bosser normalement. je sais pas a quel moment cela se déclenche et pour quelle raison 

Et je ne travaille que sur un seul exercice car je n'ai qu'un an de création de mon entreprise.

**Comment by picsou83** [comment]
Je vous confirme que c'est le message d'erreur typique d'un changement d'exercice (ou éventuellement problème de cookie web), aucune écriture n'est modifiée ou supprimée à ce stade. Il n'y a pas lieu de restaurer tout à chaque fois.

À l'occasion et sans rencontrer le problème, veuillez activer le mode dump et relever les paramètres session->{dump}, puis les copier dans ce ticket.

Paramètres => Modifier un utilisateur existant => Cliquer sur modifier => Mettre Activer le mode dump à Oui => Relever les valeurs de Paramètre session->{dump} 

exemple :
Paramètre session->{dump} 
Exercice_Cloture: 0
Exercice_debut_DMY: 01/01/2023
Exercice_debut_YMD: 2023-01-01
Exercice_fin_DMY: 31/12/2023
Exercice_fin_DMY_N1: 31/12/2022
Exercice_fin_YMD: 2023-12-31
_session_id: 2024-04-27_DGDrVEXRvQmBbYixpzFT9KdKPP
db_update_done: 1
debug: 1
dump: 1
fiscal_year: 2023
fiscal_year_offset: 0
id_client: 1
padding_zeroes: 3
preferred_datestyle: SQL, dmy
racine: base
type_compta: engagement
username: superadmin
version: 1.109

Et si jamais le problème se reproduit, veuillez relever à nouveau les paramètres session->{dump} et les copier dans ce ticket.

**Comment by picsou83** [comment]
J'ai identifié un potentiel bug lors de la modification d'une écriture. Il semble survenir lorsqu'il y a eu un nombre élevé de modifications d'écritures au cours d'une journée. Ce problème sera résolu dans la version 1.110 grâce à l'ajout d'une sécurisation pour générer un token_id unique.

**Comment by pwepwe973** [comment]
Salut
ok, c'est peut êtres cela car c'est vrais que j'avais énormément d’écriture a  ajouter et a modifier,

**Comment by picsou83** [comment]
La mise à jour 1.110, qui corrige ce problème, a été déployée.

**Comment by pwepwe973** [comment]
J’ai découvert comment reproduire le bug.

Supposons que tu sois dans l’exercice 2023 et que tu ouvres une écriture dans l’onglet XY, sans la fermer.
Ensuite, dans un autre onglet AB, tu changes d’exercice et passes de 2023 à 2024.
Puis, tu reviens dans l’exercice 2023 dans ce même onglet.
Lorsque tu retournes dans l’onglet XY, où l’écriture était ouverte, et que tu cliques sur valider, un message s’affiche : "Oups, écriture introuvable".
Le problème, c'est que dans les notes de frais, l’écriture semble toujours exister : le lien qui y pointe vers l’écriture est toujours présent, mais elle a disparu du journal des OD. Dès que tu essaies de cliquer dessus, un message s’affiche indiquant qu’elle est introuvable. En réalité, l’écriture a complètement disparu et a été définitivement détruite.

Je ne sais pas comment supprimer cette note de frais corrompue pour pouvoir recommencer à zéro, ou alors comment restaurer l’écriture perdue en lui attribuant le même numéro que celui de l’URL de l’écriture disparue.

**Comment by picsou83** [comment]
Il existe une séparation stricte entre les exercices comptables. Vous pouvez certes ouvrir plusieurs onglets simultanément, mais toujours dans le même exercice.

Si vous changez d’exercice dans un onglet (en passant de 2023 à 2024, par exemple), tous les autres onglets encore ouverts sur l’exercice précédent deviennent obsolètes. Toute interaction avec ces anciens onglets peut alors entraîner des messages d’erreur indiquant que l’écriture ou la note de frais est introuvable, car elles n’existent pas dans l’exercice actif.

Concernant votre note de frais corrompue, elle devrait normalement être accessible depuis son exercice d’origine (2023). Si ce n’est pas le cas, nous devrons examiner plus en détail ce qui s’est passé et voir s’il est possible de la restaurer ou de la supprimer proprement afin de repartir sur de bonnes bases.

Pouvez-vous confirmer si vous avez tenté d’accéder à la note de frais directement depuis l’exercice 2023 et si le problème persiste ? L’exercice en question est-il clôturé ?

**Comment by pwepwe973** [comment]
Je suis d'accord avec le fait que quand tu va dans un autre exercice que
les autres devienne obselete
Mais quand tu reviens dans l'exercice et que tu cliques sur valider que
l'écriture disparaissent ou plutôt est totalement détruit n'est pas normal
!!

 on ne l'a retrouve plus j'ai dû utiliser une ancienne sauvegarde et tout
refaire a zéro!

Et quand je parle de ne plus retrouver je parle du fait que on est dans
l'exercice en question et l'écriture a disparu totalement


Pour voir ce que je veux dire suis l'étape que je t'ai dit et tu verras.

C'est très dangereux pour les donner de l'utilisateur puisque il perd des
donné



Le jeu. 13 févr. 2025, 8:44 PM, picsou83 ***@***.***> a
écrit :

**Comment by picsou83** [comment]
Effectivement, j'ai pu reproduire le problème sur un exercice non clôturé (aucun souci constaté sur un exercice clôturé). J'ai donc apporté plusieurs modifications pour renforcer les vérifications et la sécurité, notamment lors des changements d'onglet et d'exercice :

Entry → Ajout de plusieurs sécurités pour empêcher la suppression d'une entrée si la mise à jour est incorrecte ou manquante.

Routines → Mise à jour de la routine record_staging pour s'assurer que token_id existe bien dans tbljournal_staging avant toute suppression.

Je n'avais jamais rencontré ce problème, car de mon côté, j'utilise le module Rechercher, qui permet de retrouver des écritures sur des exercices antérieurs, ainsi que de les dupliquer ou les extourner sur l'exercice en cours, sans avoir à ouvrir plusieurs onglets ni changer d'exercice.

N'hésitez pas à tester et à me faire un retour si besoin !

**Comment by picsou83** [comment]
La mise à jour 1.112, qui corrige ce problème, a été déployée.

**Comment by pwepwe973** [comment]
Salut
merci a toi pour ce taf magnifique que tu fais, problème résolu.
merci pour l'info je testerais le module que tu m'a indiqué

**Comment by pwepwe973** [comment]
Salut @picsou83, j’espère que tu vas bien !

Je réouvre ce topic pour te faire part d’un souci lié au bug qui supprimait les écritures : certaines ont bien été supprimées, mais d'autres n'ont pas été complètement supprimées, elles semblent plutôt corrompues.

Ces écritures-là sont encore visibles dans le journal, mais quand je clique dessus, j’ai le message d’erreur indiquant qu’elles sont introuvables. J’ai aussi remarqué qu’elles ont un ID égal à 0.

Tu peux le voir dans la capture d’écran que je t’ai jointe.

Du coup, j’ai prévu de te préparer une sauvegarde dans laquelle je vais supprimer toutes les écritures normales du journal et ne laisser que celle(s) corrompue(s), pour que tu puisses plus facilement identifier ce qui se passe, et peut-être voir comment les restaurer ces écriture toujours présent dans le journal mais inaccessible

Merci d’avance pour ton aide 🙏
